### PR TITLE
Replace black with ruff format and lint, run autofixes

### PR DESCRIPTION
**Pull request by martinkim0**
Per the discussion in #1002:

- Removes black formatter for Ruff formatter
- Adds Ruff lint specification

### CODE REVIEW

**Comment by martinkim0** [comment]
Most of the diff is coming from Ruff autofixes for docstrings, although there are some minor code syntax changes. It would also be ideal to use the pre-commit GitHub [app](https://github.com/apps/pre-commit-ci) or [action](https://github.com/pre-commit/action) on the repository side in case pre-commit is not run before pushing.

File: api/python/cellxgene_census/src/cellxgene_census/_release_directory.py [hunk 1]
```diff
@@ -18,37 +17,35 @@
...
-    {
-        "date": str,  # the date of retraction
-        "reason": Optional[str],  # the reason for retraction
-        "info_url": Optional[str],  # a permalink to more information
-        "replaced_by": Optional[str],  # the census version that replaces this one
-    },
-)
+
+
+class CensusLocator(TypedDict):
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/_release_directory.py:22]: this reformat loses the comments, some of which are useful/important.

File: api/python/cellxgene_census/src/cellxgene_census/_release_directory.py [hunk 2]
```diff
@@ -18,37 +17,35 @@
...
-        "flags": NotRequired[ReleaseFlags],  # flags for the release
-        "retraction": NotRequired[CensusVersionRetraction],  # if retracted, details of the retraction
-    },
-)
-CensusDirectory = Dict[CensusVersionName, Union[CensusVersionName, CensusVersionDescription]]
+ReleaseFlags = dict[ReleaseFlag, bool]
+
+
+class CensusVersionDescription(TypedDict):
+    release_date: Optional[str]
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/_release_directory.py:40]: loss of field comments

File: api/python/cellxgene_census/src/cellxgene_census/_release_directory.py [hunk 3]
```diff
@@ -72,29 +69,26 @@
...
-        "provider": Provider,  # provider of the mirror.
-        "base_uri": str,  # base URI for the mirror location, e.g. s3://cellxgene-data-public/
-        "region": Optional[str],  # region of the bucket or resource
-    },
-)
 
-CensusMirrors = Dict[CensusMirrorName, Union[CensusMirrorName, CensusMirror]]
+
+class CensusMirror(TypedDict):
+    provider: Provider
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/_release_directory.py:75]: loss of field comments

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py [hunk 4]
```diff
@@ -322,21 +319,20 @@ def get_highly_variable_genes(
     organism: str,
     measurement_name: str = "RNA",
     X_name: str = "raw",
-    obs_value_filter: Optional[str] = None,
-    obs_coords: Optional[SparseDFCoord] = None,
-    var_value_filter: Optional[str] = None,
-    var_coords: Optional[SparseDFCoord] = None,
+    obs_value_filter: str | None = None,
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:322]: AFAIK, the `|` type syntax was added in Python 3.10, but this package supports back to Python 3.8.  This seems likely to fail on Python <3.10, right?

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
We need to support Python 3.8+.   A number of the reformating changes are 3.10+

I think the `target-version` was deleted.

Note that the unit tests also caught this.

**Comment by bkmartinjr** [comment]
Unit test failure on 3.8:
```
api/python/cellxgene_census/src/cellxgene_census/_get_anndata.py:24: in <module>
    X_layers: Optional[Sequence[str]] = (),
E   TypeError: 'ABCMeta' object is not subscriptable
```

**Comment by martinkim0** [comment]
For where field comments were lost, I added them back in as docstrings. Let me know if this works for you

File: api/python/cellxgene_census/tests/experimental/pp/test_hvg.py [hunk 5]
```diff
@@ -274,18 +273,15 @@ def test_max_loess_jitter_error(small_mem_context: soma.SOMATileDBContext) -> No
...
 
         def batch_key_func(srs: pd.Series[Any]) -> str:
             raise AssertionError("should never be called without a batch key")
 
     else:
-        if isinstance(batch_key, str):
-            keys = set([batch_key])
-        else:
-            keys = set(batch_key)
+        keys = set(batch_key)
```
[the remark this answers is not in the document]
Comment (martinkim0) [reply api/python/cellxgene_census/tests/experimental/pp/test_hvg.py]: Reverted and added a noqa flag

**Comment by martinkim0** [comment]
To address
```
api/python/cellxgene_census/src/cellxgene_census/_get_anndata.py:24: in <module>
    X_layers: Optional[Sequence[str]] = (),
E   TypeError: 'ABCMeta' object is not subscriptable
```
I've replaced imports of `collections.abc.Sequence` with `typing.Sequence`.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,22 +1,6 @@
...
-        name: black-cellxgene-census
-        files: ^api/python/cellxgene_census
-        args: ["--config", "./api/python/cellxgene_census/pyproject.toml"]
-      - id: black
-        name: black-tools
-        files: ^tools
-        args: ["--config", "./tools/pyproject.toml"]
-
   - repo: https://github.com/charliermarsh/ruff-pre-commit
     rev: v0.1.15
```
[the remark this answers is not in the document]
Comment (martinkim0) [reply .pre-commit-config.yaml]: Done

File: api/python/cellxgene_census/pyproject.toml [hunk 7]
```diff
@@ -68,9 +68,65 @@ exclude = ["tests*"]  # exclude packages matching these glob patterns (empty by
 [tool.setuptools_scm]
 root = "../../.."
 
-[tool.black]
+[tool.ruff]
+src = ["api/python/cellxgene_census/src"]
 line-length = 120
-target_version = ['py39']
+extend-include = ["*.ipynb"]
+target-version = "py39"
```
Comment (martinkim0) [inline api/python/cellxgene_census/pyproject.toml:75]: Should we downgrade this to py38? Seems like a lot of the incompatible autofixes might be coming from this

Comment (bkmartinjr) [reply api/python/cellxgene_census/pyproject.toml:75]: Yes, this should be py38. Existing config is a bug.

File: api/python/cellxgene_census/src/cellxgene_census/_release_directory.py (hunk 2 above)
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/_release_directory.py:64]: still missing pre-existing comments here....

File: api/python/cellxgene_census/pyproject.toml (hunk 7 above)
Comment (bkmartinjr) [reply api/python/cellxgene_census/pyproject.toml]: note that the tools directory can use py311 as we only use that in-house

**Comment by bkmartinjr** [comment]
Another failure on 3.8 - I think this file is missing the import from `future`

```
api/python/cellxgene_census/src/cellxgene_census/_release_directory.py:60: in <module>
    ReleaseFlags = dict[ReleaseFlag, bool]
E   TypeError: 'type' object is not subscriptable
```

**Comment by martinkim0** [comment]
I'm going to go ahead and close this PR and open a new one that runs autofixes targeting py38, as there's probably going to be more errors related to this.

**Comment by bkmartinjr** [comment]
works for me.  Thank you for your perseverance.  I suspect this ended up being a bigger project than you anticipated ;-)

**Pull request closed without merging**